### PR TITLE
Improve network abstraction and transport link support ( #8 and #10 )

### PR DESCRIPTION
**Pull request by cguimaraes**
The core of zenoh-pico is abstracted from the underlying transport link, handling them in a uniform and generic way. Thus, zenoh-pico handles transport links in terms of which properties they support such as:
 - if it implements a streamed protocol or not;
 - if it implements a reliable link;

### CODE REVIEW

File: include/zenoh-pico/link/private/manager.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+#define _ZENOH_PICO_TRANSPORT_PRIVATE_LINK_MANAGER_H
+
+#include "zenoh-pico/link/result.h"
+#include "zenoh-pico/link/types.h"
+
+_zn_link_p_result_t _zn_open_link(const char *locator, const clock_t tout);
+void _zn_close_link(_zn_link_t *link);
+
+_zn_link_t *_zn_new_tcp_link(const char *s_addr, const char *port);
+_zn_link_t *_zn_new_udp_link(const char *s_addr, const char *port);
```
Comment (Mallets) [inline include/zenoh-pico/link/private/manager.h:25]: It should be called _zn_new_link_udp

File: include/zenoh-pico/link/types.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+#include "zenoh-pico/system/result.h"
+#include "zenoh-pico/system/private/unix/types.h"
+
+#define TCP_SCHEMA "tcp"
+#define UDP_SCHEMA "udp"
+
+typedef _zn_socket_result_t (*_zn_f_link_open)(void *arg, clock_t tout);
+typedef int (*_zn_f_link_close)(void *arg);
+typedef void (*_zn_f_link_release)(void *arg);
+typedef size_t (*_zn_f_link_write)(void *arg, const uint8_t *ptr, size_t len);
```
Comment (Mallets) [inline include/zenoh-pico/link/types.h:30]: Should we call `*ptr` as `*pld` to make explicit that we are talking about a payload?

File: include/zenoh-pico/link/types.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+typedef size_t (*_zn_f_link_write_all)(void *arg, const uint8_t *ptr, size_t len);
+typedef size_t (*_zn_f_link_read)(void *arg, uint8_t *ptr, size_t len);
+typedef size_t (*_zn_f_link_read_exact)(void *arg, uint8_t *ptr, size_t len);
+
+typedef struct {
+    _zn_socket_t sock;
+
+    uint8_t is_reliable;
+    uint8_t is_streamed;
```
Comment (Mallets) [inline include/zenoh-pico/link/types.h:39]: `is_reliable` and `is_streamed` could be stored in the same single `uint8_t` and save 1 byte in memory.

File: include/zenoh-pico/system/common.h [hunk 4]
```diff
@@ -57,20 +57,28 @@ time_t z_time_elapsed_ms(z_time_t *time);
...
-struct sockaddr_in *_zn_make_socket_address(const char *addr, int port);
-_zn_socket_result_t _zn_create_udp_socket(const char *addr, int port, int recv_timeout);
-
-int _zn_send_dgram_to(_zn_socket_t sock, const _z_wbuf_t *wbf, const struct sockaddr *dest, socklen_t salen);
-int _zn_recv_dgram_from(_zn_socket_t sock, _z_zbuf_t *zbf, struct sockaddr *from, socklen_t *salen);
-
-int _zn_send_wbuf(_zn_socket_t sock, const _z_wbuf_t *wbf);
-int _zn_recv_zbuf(_zn_socket_t sock, _z_zbuf_t *zbf);
-int _zn_recv_bytes(_zn_socket_t sock, uint8_t *buf, size_t len);
+// TCP
```
Comment (Mallets) [inline include/zenoh-pico/system/common.h:66]: I believe TCP and UDP (and any other) protocols should be compiled optionally through appropriate `#define` directives.

File: src/system/esp32/network.c [hunk 5]
```diff
@@ -12,271 +12,236 @@
...
-#include <net/if.h>
-#include <unistd.h>
+
 #include "zenoh-pico/system/common.h"
 #include "zenoh-pico/utils/private/logging.h"
 
-/*------------------ Interfaces and sockets ------------------*/
-char *_zn_select_scout_iface()
+/*------------------ Endpoint ------------------*/
+void _zn_release_tcp_endpoint(void *arg)
```
Comment (Mallets) [inline src/system/esp32/network.c:21]: `_zn_release_tcp_endpoint` => `_zn_release_endpoint_tcp`

File: src/system/esp32/network.c [hunk 6]
```diff
@@ -12,271 +12,236 @@
...
-    saddr->sin_port = htons(port);
+    struct addrinfo *self = (struct addrinfo*)arg;
+    freeaddrinfo(self);
+}
 
-    if (inet_pton(AF_INET, addr, &(saddr->sin_addr)) <= 0)
-    {
-        free(saddr);
+/*------------------ TCP sockets ------------------*/
+void* _zn_create_tcp_endpoint(const char *s_addr, const char *port)
```
Comment (Mallets) [inline src/system/esp32/network.c:34]: `_zn_create_tcp_endpoint` => `_zn_create_endpoint_tcp`

File: src/zenoh/transport/link/tx.c [hunk 7]
```diff
@@ -54,40 +54,41 @@ int _zn_sn_precedes(z_zint_t sn_resolution_half, z_zint_t sn_left, z_zint_t sn_r
...
-    //       in bytes of the message, resulting in the maximum length of a message being 65_535 bytes.
-    //       This is necessary in those stream-oriented transports (e.g., TCP) that do not preserve
-    //       the boundary of the serialized messages. The length is encoded as little-endian.
-    //       In any case, the length of a message must not exceed 65_535 bytes.
-    for (size_t i = 0; i < _ZN_MSG_LEN_ENC_SIZE; ++i)
-        _z_wbuf_put(buf, 0, i);
-
-    _z_wbuf_set_wpos(buf, _ZN_MSG_LEN_ENC_SIZE);
-#endif /* ZN_TRANSPORT_TCP_IP */
+    if (is_stream == 1)
```
Comment (Mallets) [inline src/zenoh/transport/link/tx.c:61]: *`is_streamed`

File: src/zenoh/transport/link/tx.c [hunk 8]
```diff
@@ -54,40 +54,41 @@ int _zn_sn_precedes(z_zint_t sn_resolution_half, z_zint_t sn_left, z_zint_t sn_r
  * Make sure that the following mutexes are locked before calling this function:
  *  - zn->mutex_tx
  */
-void __unsafe_zn_prepare_wbuf(_z_wbuf_t *buf)
+void __unsafe_zn_prepare_wbuf(_z_wbuf_t *buf, int is_stream)
```
Comment (Mallets) [inline src/zenoh/transport/link/tx.c:57]: *`is_streamed`

File: include/zenoh-pico/link/types.h (hunk 2 above)
Comment (cguimaraes) [reply include/zenoh-pico/link/types.h:30]: it is not a payload...it is the whole zenoh message (ptr will be a pointer to a zbuf)

File: include/zenoh-pico/link/types.h (hunk 3 above)
Comment (Mallets) [reply include/zenoh-pico/link/types.h:39]: Agreed

File: include/zenoh-pico/system/common.h (hunk 4 above)
Comment (Mallets) [reply include/zenoh-pico/system/common.h:66]: Agreed

File: src/system/esp32/network.c (hunk 5 above)
Comment (cguimaraes) [reply src/system/esp32/network.c:21]: Agree and done in the next commit.

File: src/system/esp32/network.c (hunk 6 above)
Comment (cguimaraes) [reply src/system/esp32/network.c:34]: Agree and done in the next commit.
I also uniformed all the other function names to follow the same naming approach.

File: src/zenoh/transport/link/tx.c (hunk 8 above)
Comment (cguimaraes) [reply src/zenoh/transport/link/tx.c:57]: Agree and done in the next commit.

**Pull request merged**